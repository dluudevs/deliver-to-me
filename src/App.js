import React, { useState, useEffect } from 'react';
import styled from 'styled-components'
import { useDispatch, useSelector } from 'react-redux'
import { ADD_RESULTS, CLEAR_RESULTS, selectRestaurant } from './slice/restaurantSlice'
import SearchByCity from './Components/SearchByCity'
import SearchByName from './Components/SearchByName'
import SortByPrice from './Components/SortByPrice'
import RestaurantItem from './Components/RestaurantItem'

const AppContainer = styled('div')(
  {
    maxWidth: '1200px',
    paddingLeft: '12px',
    paddingRight: '12px',
    margin: '0 auto'
  }
)

const SearchContainer = styled('div')(
  {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    marginBottom: '32px'
  }
)

const LoadingMessage = styled('p')(
  {
    textAlign: 'center'
  }
)

function App() {
  const dispatch = useDispatch()
  const restaurants = useSelector(selectRestaurant)
  
  const [ city, setCity ] = useState('')
  const [ name, setName ] = useState('')
  const [ priceSort, setPriceSort ] = useState('')
  
  const [ isSearching, setIsSearching ] = useState(false)
  const [ isCitySearched, setIsCitySearched ] = useState(false) 
  
  const [ filteredResults, setFilteredResults ] = useState([])

  useEffect(() => {
    const fetchRestaurants = (city) => {
      if (!city){
        return null
      }
    
      fetch(`https://opentable.herokuapp.com/api/restaurants?city=${city}&per_page=100`)
        .then(res => res.json())
        .then((data) => {
          setIsSearching(false)
          setIsCitySearched(true)
          dispatch(ADD_RESULTS(data.restaurants))
        })
    }
    
    dispatch(CLEAR_RESULTS())
    fetchRestaurants(city)
  },[city, dispatch])

  useEffect(() => {
    const regex = RegExp(name, 'gi',)
    const filtered = restaurants.filter(res => regex.test(res.name))
    setFilteredResults(filtered)
  }, [name, restaurants])

  const showResults = () => {
      // make a shallow copy of restaurants since it is read only and the sort method mutates
      let data = name ? filteredResults : [...restaurants]

      if (priceSort === 'ascending'){
        data = data.sort((a, b) => a.price - b.price)
      } else if (priceSort === 'descending'){
        data = data.sort((a, b) => b.price - a.price)
      }

      return (
        <ul style={{paddingLeft: 0}}>
          { 
            data.map((res) => <RestaurantItem key={res.id} restaurant={res}/>)
          }
        </ul>
      )
  }

  const showLoading = () => {
    if (isSearching){
      return <LoadingMessage>Searching for restaurants in: {city}</LoadingMessage>
    }

    if (isCitySearched && !isSearching){
      return <LoadingMessage>No results found for: {city}</LoadingMessage>
    }
  }

  const handleCitySubmit = (e, city) => {
    e.preventDefault()
    setCity(city)
    setIsSearching(true)
  }

  const handleFilterSubmit = (e, name) => {
    e.preventDefault()
    if (!restaurants?.length){
      return null
    }
    setName(name)
  }

  const handlePriceSort = (e) => {
    setPriceSort(e.target.value)
  }

  return (
    <AppContainer>
      <h1 style={{textAlign: 'center'}}>Deliver to Me!</h1>
      <SearchContainer>
        <SearchByCity onSubmit={handleCitySubmit} />
        <SearchByName onSubmit={handleFilterSubmit}/>
        <SortByPrice onChange={handlePriceSort}/>
      </SearchContainer>
      <div role="main">
        { restaurants?.length ? showResults() : showLoading()}
      </div>
    </AppContainer>
  )
}
export default App;
