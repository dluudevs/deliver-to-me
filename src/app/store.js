import { configureStore } from '@reduxjs/toolkit';
// import counterReducer from '../features/counter/counterSlice';
import restaurantReducer from '../slice/restaurantSlice'

export default configureStore({
  reducer: {
    restaurants: restaurantReducer,
  },
});
