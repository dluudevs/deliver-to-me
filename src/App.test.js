import React from 'react';
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux';
import store from './app/store';
import App from './App';

test('should take a snapshot', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  expect(asFragment(<App/>)).toMatchSnapshot()
});

test('displays loading message with city name', async () => {
  render(
    <Provider store={store}>
      <App />
    </Provider>
  );

  const inputValue = screen.getByLabelText('Search', {selector: 'input'})
  const [ label, submit ] = screen.getAllByText('Search')

  await userEvent.type(inputValue, 'toronto')
  await userEvent.click(submit)
  
  expect(screen.getByText('Searching for restaurants in: toronto')).toBeInTheDocument()

})

