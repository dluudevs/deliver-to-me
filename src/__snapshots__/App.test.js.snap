// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-fzoLsD etyLjA"
  >
    <h1
      style="text-align: center;"
    >
      Deliver to Me!
    </h1>
    <div
      class="sc-fzpans fMUYTq"
    >
      <form
        class="sc-AxjAm iMUxe"
      >
        <div
          class="sc-AxirZ hoemfw"
        >
          <label
            class="sc-AxiKw dsIrV"
            for="city"
          >
            Search
          </label>
          <input
            id="city"
            type="text"
            value=""
          />
        </div>
        <button
          class="sc-AxhCb gvrTmn"
          type="submit"
        >
          Search
        </button>
      </form>
      <form
        class="sc-AxjAm iMUxe"
      >
        <div
          class="sc-AxirZ hoemfw"
        >
          <label
            class="sc-AxiKw dsIrV"
            for="restaurant"
          >
            Filter
          </label>
          <input
            id="restaurant"
            type="text"
            value=""
          />
        </div>
        <button
          class="sc-AxhCb gvrTmn"
          type="submit"
        >
          Filter
        </button>
      </form>
      <div
        class="sc-AxhUy kcZXtt"
      >
        <label
          class="sc-AxgMl dHRSkO"
        >
          Sort By Price
        </label>
        <select
          id="price-sort"
          name="price"
          style="width: 169px;"
        >
          <option
            value=""
          >
            Sort Price
          </option>
          <option
            value="ascending"
          >
            Ascending
          </option>
          <option
            value="descending"
          >
            Descending
          </option>
        </select>
      </div>
    </div>
    <div
      role="main"
    />
  </div>
</DocumentFragment>
`;
